server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -419,34 +420,119 @@ public String readString() throws IOException {
             charsRef = smallSpare.get();
         }
         charsRef.length = charCount;
-        final char[] buffer = charsRef.chars;
-        for (int i = 0; i < charCount; i++) {
-            final int c = readByte() & 0xff;
-            switch (c >> 4) {
-                case 0:
-                case 1:
-                case 2:
-                case 3:
-                case 4:
-                case 5:
-                case 6:
-                case 7:
-                    buffer[i] = (char) c;
-                    break;
-                case 12:
-                case 13:
-                    buffer[i] = ((char) ((c & 0x1F) << 6 | readByte() & 0x3F));
-                    break;
-                case 14:
-                    buffer[i] = ((char) ((c & 0x0F) << 12 | (readByte() & 0x3F) << 6 | (readByte() & 0x3F) << 0));
-                    break;
-                default:
-                    throw new IOException("Invalid string; unexpected character: " + c + " hex: " + Integer.toHexString(c));
+        int charsProduced = 0;
+        int posByteArray = 0;
+        int limitByteArray = 0;
```

Comment:
I think the name `usedByteBuffer` would make more sense, I had to look at some uses to understand this was the end of the used portion of the buffer